### ISSUE TITLE
Generating a graph for files without classes?

### ISSUE THREAD

**Issue by whyboris** [Deskbot/TS-Call-Graph#2]
Hello, I tried the repository with the [example](https://github.com/Deskbot/TS-Call-Graph/blob/master/dev/examples/ClassDefinition.ts) you provide and it works great!

I have a set of TypeScript files with _no_ classes - just function calls all the way through.

_Do you think the code you have would be easy-enough to modify to work without classes?_

I'm considering trying trying to modify your code to see if I can get it to work, but if you suspect it would be rather challenging I'd love to know before I sink time into it 😅 

Specifically, I'd like to have a graph of all the calls originating from one function in a file that imports functions from other files 😁 

Thank you 🤝

**Comment by Deskbot** [comment]
I don't have any plans to develop this project further. But I would accept pull requests that don't remove functionality. I made it quickly to help me work with some code that I no longer have to work with, so that's why I only made it work within a single class body.

If you just want this quick and don't care about supporting classes:

* Use `ClassDeclarationExtractor.createDigraph` as an example.
* It should be fairly simple to find functions declared in the top level of a file and do a deep search of each for other function calls / references. You could use [a tool like this](https://github.com/Deskbot/Cognitive-Complexity-TS/blob/master/tools/what-is-it.ts) to find out what nodes are called. Be as strict as you like looking at whether the name of an outer function is shadowed by a local.
* Then input that information into an instance of `Digraph<Property>`.
* Do something with your input or with `Property` to make the types fit.
* Change where `ClassDeclarationExtractor.createDigraph` is called to call your new digraph creating function.
* Ignore / remove the class name input box.

You could make the existing codebase dual purpose and have 2 kinds of extractor and pick one based on a choice the user makes in the UI.
* And do something with `Digraph<Property>` so it can generically take functions and methods. You could make `privacy` optional in `NodeInput`. There are some redundant attributes in `Property`. But there would be some type wrangling to make that possible.

You could try to combine them together but that seems quite hard. If you have methods/functions with the same name, how will you know which one is called?

**Comment by Deskbot** [comment]
To make it work across multiple files means doing a lot of work that I haven't done.

Maybe the core of what you want, if you don't want to spend ages figuring out how d3 works or how to do a UI, is just "draw.ts".

**Comment by whyboris** [comment]
Thank you very much for the detailed instructions! 

I spent some time working with it, but realized I may have an easier time (and learn a lot too) if I create something from scratch (borrowing some of your ideas / code).

I created my own [TypeScript-Call-Graph](https://github.com/whyboris/TypeScript-Call-Graph) repository for a *CLI* tool that will hopefully do something similar to what you've already accomplished 😅 

I tried numerous graphing libraries but most were annoying to use, so I decided to work with D3 (while it's super-powerful, I'll have to learn a lot first and then micromanage things). Hope I can get it to work the way I'd like 🤞
